File: lib/features/cart/data/cart_controller.dart
import 'package:flutter/foundation.dart';

import '../../../core/services/foodly_api_service.dart';

class CartController extends ChangeNotifier {
  CartController({required this.api});

  final FoodlyApiService api;
  Map<String, dynamic> _cart = const {'items': [], 'subtotal': 0, 'deliveryFee': 0, 'total': 0};
  bool _isLoading = false;
  String? _errorMessage;

  Map<String, dynamic> get cart => _cart;
  List<dynamic> get items => _cart['items'] as List<dynamic>? ?? const [];
  int get itemCount => items.fold(0, (sum, item) => sum + ((item as Map<String, dynamic>)['quantity'] as int));
  bool get isLoading => _isLoading;
  String? get errorMessage => _errorMessage;

  Future<void> load() async {
    _isLoading = true;
    _errorMessage = null;
    notifyListeners();
    try {
      _cart = await api.getCart();
    } on FoodlyApiException catch (error) {
      _errorMessage = error.message;
    } finally {
      _isLoading = false;
      notifyListeners();
    }
  }

  Future<String?> addItem(String foodItemId, {int quantity = 1}) => _change(
    () => api.addCartItem(foodItemId: foodItemId, quantity: quantity),
  );

  Future<String?> updateItem(String foodItemId, int quantity) => _change(
    () => api.updateCartItem(foodItemId: foodItemId, quantity: quantity),
  );

  Future<String?> removeItem(String foodItemId) => _change(() => api.removeCartItem(foodItemId));

  Future<String?> clear() => _change(api.clearCart);

  Future<Map<String, dynamic>?> createOrder({required String deliveryAddress}) async {
    _isLoading = true;
    _errorMessage = null;
    notifyListeners();
    try {
      final order = await api.createOrder(deliveryAddress: deliveryAddress);
      _cart = const {'items': [], 'subtotal': 0, 'deliveryFee': 0, 'total': 0};
      return order;
    } on FoodlyApiException catch (error) {
      _errorMessage = error.message;
      return null;
    } finally {
      _isLoading = false;
      notifyListeners();
    }
  }

  Future<String?> _change(Future<Map<String, dynamic>> Function() change) async {
    _isLoading = true;
    _errorMessage = null;
    notifyListeners();
    try {
      _cart = await change();
      return null;
    } on FoodlyApiException catch (error) {
      _errorMessage = error.message;
      return error.message;
    } finally {
      _isLoading = false;
      notifyListeners();
    }
  }
}

File: backend/src/controllers/cart_controller.js
import { addCartItem, clearCart, getCartDetails, removeCartItem, updateCartItem } from '../services/cart_service.js';
import { asyncHandler } from '../utils/async_handler.js';

export const getCart = asyncHandler(async (request, response) => {
  response.status(200).json({ success: true, data: await getCartDetails(request.user._id) });
});

export const addItem = asyncHandler(async (request, response) => {
  const { foodItemId, quantity = 1 } = request.body;
  response.status(200).json({ success: true, data: await addCartItem(request.user._id, foodItemId, quantity) });
});

export const updateItem = asyncHandler(async (request, response) => {
  response.status(200).json({ success: true, data: await updateCartItem(request.user._id, request.params.foodItemId, request.body.quantity) });
});

export const removeItem = asyncHandler(async (request, response) => {
  response.status(200).json({ success: true, data: await removeCartItem(request.user._id, request.params.foodItemId) });
});

export const clear = asyncHandler(async (request, response) => {
  response.status(200).json({ success: true, data: await clearCart(request.user._id) });
});
